src/In.ProjectEKA.HipService/Discovery/PatientDiscovery.cs

```diff
@@ -1,39 +1,78 @@
 namespace In.ProjectEKA.HipService.Discovery
 {
     using System;
+    using System.Collections.Generic;
     using System.Linq;
     using System.Threading.Tasks;
     using HipLibrary.Patient;
     using HipLibrary.Patient.Model;
+    using In.ProjectEKA.HipService.Link;
+    using In.ProjectEKA.HipService.Link.Model;
 
     public class PatientDiscovery : IDiscovery
     {
         private readonly Filter filter;
         private readonly IMatchingRepository matchingRepository;
         private readonly IDiscoveryRequestRepository discoveryRequestRepository;
+        private readonly ILinkPatientRepository linkPatientRepository;
+        private readonly IPatientRepository patientRepository;
 
         public PatientDiscovery(
-            IMatchingRepository patientRepository,
-            IDiscoveryRequestRepository discoveryRequestRepository)
+            IMatchingRepository matchingRepository,
+            IDiscoveryRequestRepository discoveryRequestRepository,
+            ILinkPatientRepository linkPatientRepository,
+            IPatientRepository patientRepository)
         {
-            matchingRepository = patientRepository;
+            this.matchingRepository = matchingRepository;
             this.discoveryRequestRepository = discoveryRequestRepository;
+            this.linkPatientRepository = linkPatientRepository;
+            this.patientRepository = patientRepository;
             filter = new Filter();
         }
 
         public async Task<Tuple<DiscoveryRepresentation, ErrorRepresentation>> PatientFor(DiscoveryRequest request)
         {
-            var patients = await matchingRepository.Where(request);
-            var (patient, error) = DiscoveryUseCase.DiscoverPatient(filter.Do(patients, request).AsQueryable());
+            var (linkRequest, exception) = await linkPatientRepository.GetLinkedCareContexts(request.Patient.Id);
+            if (exception != null)
+            {
+                return new Tuple<DiscoveryRepresentation, ErrorRepresentation>(null,
+                    new ErrorRepresentation(new Error(ErrorCode.FailedToGetLinkedCareContexts,
+                        "Failed to get Linked Care Contexts")));
+            }
+
+            if (linkRequest != null)
+            {
+                return patientRepository.PatientWith(linkRequest.PatientReferenceNumber)
+                    .Map(patient => new Tuple<DiscoveryRepresentation, ErrorRepresentation>(
+                        new DiscoveryRepresentation(patient.ToPatientEnquiryRepresentation(
+                            GetUnlinkedCareContexts(linkRequest, patient))),
+                        null)
+                    ).ValueOr(
+                        new Tuple<DiscoveryRepresentation, ErrorRepresentation>(
+                            null,
+                            new ErrorRepresentation(new Error(ErrorCode.NoPatientFound, ErrorMessage.NoPatientFound))));
+            }

```

Comment:
We have to log this scenario, since we found one link entry in HIP but there's no such patient.

Reply:
Sure.